### ISSUE TITLE
kits with very large memberships >1200 extremely slow to load

### ISSUE THREAD

**Issue by joliet** [DukeLearningInnovation/kits#606]
**Describe the bug**
Kits with very large memberships extremely slow to load.

**To Reproduce**
1. Go to kits.duke.edu
2. Admin Console
3. Click Search All Kits
4. Search for kit 'Econ Major Orientation'
5. Click View/Edit Roster for kit
6. Kit takes about 45 sec to load in View Mode
7. Clicking on Edit mode takes another 45 sec

**Expected behavior**
I would expect the membership list to show more quickly (10 seconds or less?).  If this is not possible, we should probably show a loading or progress image to provide feedback to users.

**Screenshots**

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Additional context**

**Comment by ewendland23** [comment]
@joliet - Can you give a few examples of sites where you have seen this?  Thanks!

**Comment by joliet** [comment]
In addition to the 'Econ Major Orientation' (kit id 16510) that I mentioned in the steps to replicate there's: divnity_writing_ctr (~20 sec to load); Student Commons MSN (kit id 22760) took about 1 min 2 sec to load.

Let me know if you need more examples.

**Comment by ewendland23** [comment]
I have deleted the 1400! people from Econ Major.  I will work on divinity_writing_ctr next.

**Comment by chrislorch** [comment]
Another example with 1650 users: [Outlook Technical Support](https://sakai.duke.edu/portal/site/bdc8640e-0474-4d13-9427-862cd4ee6f6f/)

**Comment by joliet** [comment]
@noahhh I think we discussed adding pagination to the member list in a kit to speed this up.  Is that right?  This would be a good issue to work on soon.  Reach out to me on Teams about when you think you might have time to work on this.

**Comment by noahhh** [comment]
@joliet is there a way we can replicate this in test so we can verify any changes I make before going to prod?

**Comment by joliet** [comment]
@noahhh I was just trying to add ~1500 members to my kit TEAM 2021-02 My test team (kit id 2184) so that we could use that one but I'm having trouble.  At first I made a mistake and used netid@duke.edu in the NetID option.  Kits gave me an error.  I stripped out the @duke.edu and tried again with just NetID but nothing happened.  I waited maybe 10 minutes for feedback to to see the users in the list with no luck.  I tried again with email and still no response from Kits.  Can you see what might be happening there?

**Comment by noahhh** [comment]
@joliet I didn't see any errors in the logs (I believe I looked in the right place).  When I visit the kit in test, I can see all of the members.  Are you still unable to see them?

**Comment by joliet** [comment]
@noahhh Oh weird.  The bulk add page where I added them does not show them so I didn't know the process finished.

**Comment by joliet** [comment]
Hi @noahhh .  I was testing this kit and noticed a couple of issue with search.  I started researching that and now that I've come back to Kits for more testing, I can't seem to find my kit with the large member list - Team:2021-02:2184.  I don't see in My Kits.  Did you delete that?  I couldn't find any evidence of deletion in the activity logs.

**Comment by noahhh** [comment]
@joliet Huh, odd.  I haven't removed it, but I also don't see it anymore.  Not sure where it went, as I don't see it in the database either.

**Comment by ewendland23** [comment]
Oh! That might have been me @noahhh and @joliet.  I was cleaning up all my various Kits on test yesterday.  I had so many I just cleaned house.  I am sorry I removed your test.  I wish the names were something like "Big Test case.  Do not delete"

**Comment by joliet** [comment]
Thanks for letting us know @ewendland23

**Comment by noahhh** [comment]
@joliet I made a change to the pagination in test.  Can you play around with it when you get a chance?

I have not added pagination to combined kits, as I'm not sure how that would look/work because we display all of the rosters.  Maybe we just tell people not to combine large rosters (or hopefully the people with large rosters won't need to combine them anyway)?

**Comment by joliet** [comment]
You were able to format it.  Nice!  Looks like sorting only works with the current page, is that right?  Searching seems to work fine.  I don't think you have to worry about paginating combined kits.  To my knowledge, there are no Duke courses with enrollments that are over 1K (probably not even over 200).  I will test this a bit more but so far, looks good to me!

**Comment by noahhh** [comment]
@joliet sorting is a javascript based sort, only for what's on the page (it's how we had it  before pagination).  I can change it to sort everything if you'd like, it would make sense that way.  That would cause a page refresh to get the sorted columns.

**Comment by joliet** [comment]
@noahhh that would be great - if you changed it to sort everything.  I think it's worth the page refresh.  That could save time for any folks who actually want to navigate clicking through the pages.

**Comment by noahhh** [comment]
@joliet I added some sorting changes to test

**Comment by joliet** [comment]
Hi @noahhh .  This looks good!  Everything seems to be working fine.  Jen and I are discussing the Search box design.  We've got some inconsistencies with Search in Kits which I've outlined here - https://www.figma.com/file/fnBslhg8ImBlVjlGn0n8Lf/Search-Design-in-Kits?node-id=0%3A1

We definitely want an X to clear results, but haven't landed on a decision yet for a change with the search icon placement and color.  Hoping to wrap this discussion up soon and will create a new GH issue to replace Search on pages that need conforming.

**Comment by joliet** [comment]
@noahhh And Jen's design is done. : ) Here's a link to the Figma file that contains the Kits pattern library.  The Search design is on the last page - https://www.figma.com/file/dB483dx21JgjFYbJD059Dr/Pattern-Library---Kits?node-id=0%3A1027

If you could add this to the Add Members page, we can wrap up this issue.  I will create a new issue for replacing all other Search designs with this new one.

**Comment by noahhh** [comment]
@joliet should the 'clear' (X) just clear the input box and they have to search on an empty input to return to normal, or should it reload the page with no query (so everyone shows back up)?

**Comment by noahhh** [comment]
@joliet I put some changes for the search bar out in test

**Comment by joliet** [comment]
Hi @noahhh .  The X should just clear the input box allowing new search terms to be entered.  I just did some testing with other search boxes and this was the consistent behavior.

**Comment by joliet** [comment]
Oh, I see your new comment.  Thanks!

**Comment by joliet** [comment]
Hi @noahhh  I just took a look at this.  Are you still working on it?  The search doesn't seem to be working and the X goes away after you type in search term in the box and click the magnifying glass icon.  No worries if you are still working on this.  I just didn't want to leave you waiting on me to test.

**Comment by noahhh** [comment]
@joliet Sorry, this was unfinished.  I just pushed up some new code for this for you to test out.

**Comment by joliet** [comment]
No worries @noahhh !  I hope you had a great two days off!

I just tested your changes and it looks like deleting a user from the search results breaks the pagination.  It comes back after a refresh but it would be better if it the pagination would stick after the snackbar confirmation that the member was removed.

**Comment by joliet** [comment]
This looks good to me @noahhh !  We can move this to prod whenever you and Liz think it's ok.
